go/pools/fast_pool.go

```diff
@@ -0,0 +1,627 @@
+// in the future through the SetCapacity function.
+//
+// If a resource is unused beyond idleTimeout, it's discarded.
+// An idleTimeout of 0 means that there is no timeout.
+//
+// minActive is used to prepare and maintain a minimum amount
+// of active resources. Any errors when instantiating the factory
+// will cause the active resource count to be lower than requested.
+func NewFastPool(factory CreateFactory, capacity, maxCap int, idleTimeout time.Duration, minActive int) *FastPool {
+	if capacity <= 0 || maxCap <= 0 || capacity > maxCap {
+		panic(errors.New("invalid/out of range capacity"))
+	}
+	if minActive > capacity {
+		panic(fmt.Errorf("minActive %v higher than capacity %v", minActive, capacity))
+	}
+
+	p := &FastPool{
+		factory: factory,
+		pool:    make(chan resourceWrapper, maxCap),
+	}
+
+	p.state = State{
+		Capacity:  capacity,
+		MinActive: minActive,
+	}
+
+	p.ensureMinActive()
+	p.SetIdleTimeout(idleTimeout)
+
+	go p.closeIdleResources()
+	go p.maintainMinActive()
+
+	return p
+}
+
+// create a resource wrapper. return an error if the factory function failed.
+func (p *FastPool) create() (resourceWrapper, error) {
+	r, err := p.factory()
+	if err != nil {
+		return resourceWrapper{}, err
+	}
+
+	return resourceWrapper{
+		resource: r,
+		timeUsed: time.Now(),
+	}, nil
+}
+
+// createType is used by safeCreate to know which counter to increment.
+type createType int
+
+const (
+	forUse createType = iota
+	forPool
+)
+
+// safeCreate will prevent allocating a resource past the capacity of the pool.
+func (p *FastPool) safeCreate(ct createType) (resourceWrapper, error) {
+	var capacity bool
+	p.withLock(func() {
+		capacity = p.hasFreeCapacity()
+		p.state.Spawning++
+	})
+
+	if !capacity {
+		p.withLock(func() {
+			p.state.Spawning--
+		})
+		return resourceWrapper{}, errNeedToQueue
+	}
+
+	wrapper, err := p.create()
+	if err != nil {
+		p.withLock(func() {
+			p.state.Spawning--
+		})
+		return resourceWrapper{}, err
+	}
+
+	p.withLock(func() {
+		switch ct {
+		case forUse:
+			p.state.InUse++
+		case forPool:
+			p.state.InPool++
+		}
+		p.state.Spawning--
+	})
+
+	return wrapper, nil
+}
+
+// maintainMinActive will keep calling ensureMinActive until quit.
+// This function is intended to be used as a goroutine.
+func (p *FastPool) maintainMinActive() {
+	for !p.IsClosed() {
+		p.ensureMinActive()
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+// ensureMinActive keeps at least a certain amount of resources instantiated.
+func (p *FastPool) ensureMinActive() {
+	p.Lock()
+	if p.state.MinActive == 0 || p.state.Closed {
+		p.Unlock()
+		return
+	}
+	required := p.state.MinActive - p.active()
+	p.Unlock()
+
+	for i := 0; i < required; i++ {
+		r, err := p.safeCreate(forPool)
+		if err == errNeedToQueue {
+			// Not enough room in the pool. Aborting.
+			return
+		} else if err != nil {
+			// TODO(gak): How to handle factory error not initiated by the user?
+			return
+		}
+
+		select {
+		case p.pool <- r:
```

Comment:
If the pool is closed just before reaching this point, this will panic. It appears that Close cannot succeed while this function is active. But I see holes in safeCreate where locks are acquired and released which may give opportunity for Close to succeed while this loop is executing.